src/slow_log/slow_log.c

```diff
@@ -0,0 +1,224 @@
+}
+
+static SlowLog *_SlowLog_Create(int thread_count) {
+	SlowLog *slowlog = rm_malloc(sizeof(SlowLog));
+	slowlog->count = thread_count;
+	slowlog->lookup = rm_malloc(sizeof(rax *) * thread_count);
+	slowlog->min_heap = rm_malloc(sizeof(heap_t *) * thread_count);
+	slowlog->locks = rm_malloc(sizeof(pthread_mutex_t) * thread_count);
+
+	for(int i = 0; i < thread_count; i++) {
+		slowlog->lookup[i] = raxNew();
+		slowlog->min_heap[i] = heap_new(_slowlog_elem_compare, NULL);
+		assert(pthread_mutex_init(slowlog->locks + i, NULL) == 0);
+	}
+
+	return slowlog;
+}
+
+static void _SlowLog_Free(SlowLog *slowlog) {
+	assert(slowlog);
+
+	for(int i = 0; i < slowlog->count; i++) {
+		rax *lookup = slowlog->lookup[i];
+		heap_t *heap = slowlog->min_heap[i];
+
+		/* Free each slowlog item, these are shared between
+		 * the rax and the heap objects. */
+		while(heap_count(heap)) {
+			SlowLogItem *item = heap_poll(heap);
+			_SlowLog_Item_Free(item);
+		}
+
+		heap_free(heap);
+		raxFree(lookup);
+		assert(pthread_mutex_destroy(slowlog->locks + i) == 0);
+	}
+
+	// Free arrays.
+	rm_free(slowlog->locks);
+	rm_free(slowlog->lookup);
+	rm_free(slowlog->min_heap);
+	rm_free(slowlog);
+	slowlog = NULL;
+}
+
+static void _SlowLog_Add(SlowLog *slowlog, int t_id, const char *cmd, const char *graph_id,
+						 const char *query, double latency) {
+	assert(graph_id && query && latency > 0);
+
+	char *s;
+	rax *lookup = slowlog->lookup[t_id];
+	heap_t *heap = slowlog->min_heap[t_id];
+	size_t len = _compute_key(&s, cmd, graph_id, query);
+	pthread_mutex_lock(slowlog->locks + t_id);
+
+	SlowLogItem *existing_item = (SlowLogItem *)raxFind(lookup, (unsigned char *)s, strlen(s));
+
+	if(existing_item != raxNotFound) {
+		/* A similar item is already part of the slowlog.
+		 * see if we need to replace it. */
+		if(existing_item->latency < latency) {
+
+			// Replace existing item with new item.
+			SlowLogItem *item = _SlowLogItem_New(cmd, graph_id, query, latency);
+
+			// Overwrite item in lookup.
+			raxInsert(lookup, (unsigned char *)s, len, item, NULL);
+
+			// Remove existing item from heap.
+			// assert(heap_remove_item(heap, existing_item) == existing_item);
+			assert(heap_remove_item(heap, existing_item) != NULL);
+			_SlowLog_Item_Free(existing_item);
+
+			// Introduce new item.
+			heap_offer(slowlog->min_heap + t_id, item);
+		}
+
+		/* Done, if an item already existed we either replace it
+		 * or leave it. */
+		goto cleanup;
+	}
+
+	/* Similar item does not exist in the log.
+	 * Check if there's enough room to store item. */
+	int introduce_item = 0;
+	if(heap_count(heap) < SLOW_LOG_SIZE) {
+		introduce_item = 1;
+	} else {
+		// Not enough room, see if item should be tracked.
+		SlowLogItem *top = heap_peek(heap);
+		if(top->latency < latency) {
+			top = heap_poll(heap);
+
+			char *top_str;
+			size_t len = SlowLogItem_ToString(top, &top_str);
+			assert(raxRemove(lookup, (unsigned char *)top_str, len, NULL) == 1);
```

Comment:
I get a number of crashes on these calls in the multi-threaded stress tests, indicating that this code isn't thread-safe.

One sample crash occurs when the item `top` retrieved from a valid-looking heap is full of invalid memory:
```
(gdb) p *top
$5 = {
  cmd = 0x3432313000000020 <error: Cannot access memory at address 0x3432313000000020>, 
  time = 140736783809344, 
  query = 0x7fffd6017020 "\030", 
  graph_id = 0x7fffd600ea50 "\020", 
  latency = 6.9533209954723707e-310
}
```
This will, depending on contents, crash in `top_str` construction or trigger the assertion failure on trying to remove invalid inputs.